Suggest: Retain entered value on unfocus

#### Environment

- __Package version(s)__: `@blueprintjs/select 3.2.0`
- __Browser and OS versions__: any

#### Feature request

I'm trying to make`Suggest` work for search suggestions where users can enter any value and be provided with search suggestions, very similar to Google's search box.

My issue is that `Suggest` discards any non-matching user-entered value on unfocus. If no value is selected, it clears the entered value on unfocus. If a value has been previously selected, the entered  value is replaced with the previously selected value.

#### Examples

I'd suggest a new option `keepEnteredValues` option to retain non-matching values or alternatively, make it the default behaviour, like it is for `MultiSelect`.

Not sure I can actually contribute as I'm failing to get a development environment up.

`yarn run compile` errors with `TS2694: Namespace 'ts' has no exported member 'ImportTypeNode'`
`yarn dev:select` spews a lot of errors like `TS2307: Cannot find module './generated/iconNames'`

(This is using node 10.12, yarn 1.10.1, macOS 10.14)

My basic idea was to drop [this line](https://github.com/palantir/blueprint/blob/15f8a82ebaa3afff52f14ea2ae09d433b2d9bedf/packages/select/src/components/select/suggest.tsx#L177) and debug what's happening to [this value](https://github.com/palantir/blueprint/blob/15f8a82ebaa3afff52f14ea2ae09d433b2d9bedf/packages/select/src/components/select/suggest.tsx#L156).

--- a/packages/select/src/components/select/suggest.tsx
+++ b/packages/select/src/components/select/suggest.tsx
@@ -80,8 +80,8 @@ export class Suggest<T> extends React.PureComponent<ISuggestProps<T>, ISuggestSt
     }
 
     private TypedQueryList = QueryList.ofType<T>();
-    private input?: HTMLInputElement | null;
-    private queryList?: QueryList<T> | null;
+    private input: HTMLInputElement | null = null;
+    private queryList: QueryList<T> | null = null;
     private refHandlers = {
         input: (ref: HTMLInputElement | null) => {
             this.input = ref;
@@ -133,6 +133,11 @@ export class Suggest<T> extends React.PureComponent<ISuggestProps<T>, ISuggestSt
         const { placeholder = "Search..." } = inputProps;
 
         const selectedItemText = selectedItem ? this.props.inputValueRenderer(selectedItem) : "";
+        // placeholder shows selected item while open.
+        const inputPlaceholder = isOpen && selectedItemText ? selectedItemText : placeholder;
+        // value shows query when open, and query remains when closed if nothing is selected.
+        const inputValue = isOpen ? listProps.query : selectedItemText || listProps.query;
+
         return (
             <Popover
                 autoFocus={false}
@@ -147,13 +152,13 @@ export class Suggest<T> extends React.PureComponent<ISuggestProps<T>, ISuggestSt
             >
                 <InputGroup
                     {...inputProps}
-                    placeholder={isOpen && selectedItemText ? selectedItemText : placeholder}
+                    placeholder={inputPlaceholder}
                     inputRef={this.refHandlers.input}
                     onChange={listProps.handleQueryChange}
                     onFocus={this.handleInputFocus}
                     onKeyDown={this.getTargetKeyDownHandler(handleKeyDown)}
                     onKeyUp={this.getTargetKeyUpHandler(handleKeyUp)}
-                    value={isOpen ? listProps.query : selectedItemText}
+                    value={inputValue}
                 />
                 <div onKeyDown={handleKeyDown} onKeyUp={handleKeyUp}>
                     {listProps.itemList}

